src/client/Microsoft.Identity.Client/ManagedIdentity/CredentialSource.cs

```diff
@@ -0,0 +1,185 @@
+                string baseUrl = "https://centraluseuap.mtlsauth.microsoft.com/";
+                string tenantId = credentialResponse.TenantId;
+                string tokenEndpoint = "/oauth2/v2.0/token?slice=testslice";
+                Uri url = new(string.Join("", baseUrl, tenantId, tokenEndpoint));
+
+                ManagedIdentityRequest request = new(HttpMethod.Post, url);
+                var scope = parameters.Resource + "/.default";
+                request.Headers.Add("x-ms-client-request-id", _requestContext.CorrelationId.ToString("D"));
+                request.BodyParameters.Add("grant_type", "client_credentials");
+                request.BodyParameters.Add("client_id", credentialResponse.ClientId);
+                request.BodyParameters.Add("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer");
+                request.BodyParameters.Add("client_assertion", credentialResponse.Credential);
+                request.BodyParameters.Add("scope", scope);
+                request.BindingCertificate = _requestAssertionInfo.BindingCertificate;
+
+                var mergedclaims = ClaimsHelper.GetMergedClaimsAndClientCapabilities(
+                    parameters.Claims, _requestContext.ServiceBundle.Config.ClientCapabilities);
+
+                request.BodyParameters.Add("claims", mergedclaims);
+
+                _requestContext.Logger.Verbose(() => $"[Managed Identity] Sending token request to mtls " +
+                $"endpoint : {credentialResponse.RegionalTokenUrl}.");
+
+                response = await _requestContext.ServiceBundle.HttpManager
+                        .SendPostForceResponseAsync(
+                            request.ComputeUri(),
+                            request.Headers,
+                            request.BodyParameters,
+                            request.BindingCertificate,
+                            _requestContext.Logger, cancellationToken: cancellationToken).ConfigureAwait(false);
+            }
+
+            return await base.HandleResponseAsync(parameters, response, cancellationToken).ConfigureAwait(false);
+        }
+
+        private CredentialResponse GetCredentialResponse(HttpResponse response)
+        {
+            CredentialResponse credentialResponse = JsonHelper.DeserializeFromJson<CredentialResponse>(response?.Body);
+
+            if (credentialResponse == null || credentialResponse.Credential.IsNullOrEmpty())
+            {
+                _requestContext.Logger.Error("[Managed Identity] Response is either null or insufficient for authentication.");
+                throw new MsalManagedIdentityException(
+                    MsalError.ManagedIdentityRequestFailed,
+                    MsalErrorMessage.ManagedIdentityInvalidResponse,
+                    _sourceType);
+            }
+
+            return credentialResponse;
+        }
+
+        private static string CreateCredentialPayload(X509Certificate2 x509Certificate2)
+        {
```

Comment:
HOw expensive is this? Please add more logging and use the `LogBlockDuration` to have some stats.